provider/queue.go

```diff
@@ -0,0 +1,181 @@
+	// e.g. provider vs reprovider
+	name    string
+	ctx     context.Context
+	tail    uint64
+	head    uint64
+	ds      datastore.Datastore // Must be threadsafe
+	dequeue chan cid.Cid
+	enqueue chan cid.Cid
+}
+
+// NewQueue creates a queue for cids
+func NewQueue(ctx context.Context, name string, ds datastore.Datastore) (*Queue, error) {
+	namespaced := namespace.Wrap(ds, datastore.NewKey("/"+name+"/queue/"))
+	head, tail, err := getQueueHeadTail(ctx, name, namespaced)
+	if err != nil {
+		return nil, err
+	}
+	q := &Queue{
+		name:    name,
+		ctx:     ctx,
+		head:    head,
+		tail:    tail,
+		ds:      namespaced,
+		dequeue: make(chan cid.Cid),
+		enqueue: make(chan cid.Cid),
+	}
+	q.work()
+	return q, nil
+}
+
+// Enqueue puts a cid in the queue
+func (q *Queue) Enqueue(cid cid.Cid) {
+	select {
+	case q.enqueue <- cid:
+	case <-q.ctx.Done():
+	}
+}
+
+// Dequeue returns a channel that if listened to will remove entries from the queue
+func (q *Queue) Dequeue() <-chan cid.Cid {
+	return q.dequeue
+}
+
+type entry struct {
+	cid cid.Cid
+	key datastore.Key
+}
+
+// Look for next Cid in the queue and return it. Skip over gaps and mangled data
+func (q *Queue) nextEntry() (datastore.Key, cid.Cid) {
+	for {
+		if q.head >= q.tail {
+			return datastore.Key{}, cid.Undef
+		}
+
+		key := q.queueKey(q.head)
+		value, err := q.ds.Get(key)
+
+		if err == datastore.ErrNotFound {
+			log.Warningf("Error missing entry in queue: %s", key)
+			q.head++ // move on
+			continue
+		} else if err != nil {
+			log.Warningf("Error fetching from queue: %s", err)
+			continue
+		}
+
+		c, err := cid.Parse(value)
+		if err != nil {
+			log.Warningf("Error marshalling Cid from queue: ", err)
+			q.head++
+			err = q.ds.Delete(key)
+			continue
+		}
+
+		return key, c
+	}
+}
+
+// Run dequeues and enqueues when available.
+func (q *Queue) work() {
+	go func() {
+		var k datastore.Key = datastore.Key{}
+		var c cid.Cid = cid.Undef
+
+		for {
+			if c == cid.Undef {
+				k, c = q.nextEntry()
+			}
+
+			// If c != cid.Undef set dequeue and attempt write, otherwise wait for enqueue
+			var dequeue chan cid.Cid
+			if c != cid.Undef {
+				dequeue = q.dequeue
+			}
+
+			select {
+			case toQueue := <-q.enqueue:
+				nextKey := q.queueKey(q.tail)
+
+				if err := q.ds.Put(nextKey, toQueue.Bytes()); err != nil {
+					log.Errorf("Failed to enqueue cid: %s", err)
+					continue
+				}
+
+				q.tail++
+			case dequeue <- c:
+				err := q.ds.Delete(k)
+
+				if err != nil {
+					log.Errorf("Failed to delete queued cid %s with key %s: %s", c, k, err)
+					continue
+				}
+				c = cid.Undef
+				q.head++
+			case <-q.ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (q *Queue) queueKey(id uint64) datastore.Key {
+	return datastore.NewKey(strconv.FormatUint(id, 10))
+}
+
+// crawl over the queue entries to find the head and tail
+func getQueueHeadTail(ctx context.Context, name string, datastore datastore.Datastore) (uint64, uint64, error) {
+	q := query.Query{}
```

Comment:
So, there's a _much_ faster way to do this. We can:

1. Set `Query.Prefix` to our key's prefix. That'll ensure that we only crawl over that prefix.
2. We can set `Query.Order = []query.Order{query.OrderByKey}` and then pull off the first item to get the head.
3. We can set a limit of 1 (I think?).

Note: We'll have to encode our numbers with a fixed-width (e.g., 16 hex characters) to make this work as we use lexical ordering.

Reply:
(we can perform a second query to find the tail)

Reply:
@Stebalien I want to get a quick priority check on this given the fact that we need to add features to the query mechanism for mount in datastore. (see https://github.com/ipfs/go-datastore/pull/121). I don't mind continuing that work, but I wonder, can we get this work merged and out while working on the datastore (or whatever is the most important thing to do next) and patch a more performant solution when the prerequisite work is complete?

Reply:
If we set the query _prefix_ so we only have to walk over the records in the provider queue, that should be enough for now.

Reply:
@Stebalien we actually have the whole datastore namespaced and that's why the query doesn't have a prefix. My understanding is that this will have the affect you're looking for. Is that true?

Reply:
Got it. Then the only thing left on this PR is docs around Provide/Run (sorry, being picky about this is one of _my_ KRs).

Reply:
Don't be sorry, I appreciate it and I'm working on those now. I just focused on the bigger issue to get it resolved ASAP.